library/ssl_tls13_generic.c

```diff
@@ -1714,24 +1736,47 @@ int mbedtls_ssl_tls13_parse_record_size_limit_ext(mbedtls_ssl_context *ssl,
 
     MBEDTLS_SSL_DEBUG_MSG(2, ("RecordSizeLimit: %u Bytes", record_size_limit));
 
-    /* RFC 8449, section 4:
-     *
-     * Endpoints MUST NOT send a "record_size_limit" extension with a value
-     * smaller than 64.  An endpoint MUST treat receipt of a smaller value
-     * as a fatal error and generate an "illegal_parameter" alert.
-     */
-    if (record_size_limit < MBEDTLS_SSL_RECORD_SIZE_LIMIT_MIN) {
-        MBEDTLS_SSL_PEND_FATAL_ALERT(
-            MBEDTLS_SSL_ALERT_MSG_ILLEGAL_PARAMETER,
-            MBEDTLS_ERR_SSL_ILLEGAL_PARAMETER);
-        return MBEDTLS_ERR_SSL_ILLEGAL_PARAMETER;
+    ret = mbedtls_ssl_is_record_size_limit_valid(ssl, record_size_limit);
+    if (ret != 0) {
+        return ret;
     }
 
     ssl->session_negotiate->record_size_limit = record_size_limit;
 
     return 0;
 }
 
+MBEDTLS_CHECK_RETURN_CRITICAL
+int mbedtls_ssl_tls13_write_record_size_limit_ext(mbedtls_ssl_context *ssl,
+                                                  uint16_t record_size_limit,
+                                                  unsigned char *buf,
+                                                  const unsigned char *end,
+                                                  size_t *out_len)
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    unsigned char *p = buf;
+    *out_len = 0;
+
+    ret = mbedtls_ssl_is_record_size_limit_valid(ssl, record_size_limit);
+    if (ret != 0) {
+        return ret;
+    }
```

Comment:
In that case the error is a configuration (build time one currently) thus I do not think we should send an MBEDTLS_SSL_ALERT_MSG_ILLEGAL_PARAMETER alert (especially in the client case where we are currently building the ClientHello message) and thus we cannot use `mbedtls_ssl_is_record_size_limit_valid()` as it is here.

I do not know if we will have to but if at some point we add an `mbedtls_ssl_conf_record_size_limit()` function (similar to `mbedtls_ssl_conf_max_frag_len()`) I'd expect that kind of check to be done in this function and not later on when we use it and the check here that the `record_size_limit` is large enough would be redundant. Furthermore, with a configuration function, the `record_size_limit` would not be passed as a parameter to the
function but taken from `ssl->conf->record_size_limit`.

Thus all together I propose:
1) To change the signature of this function to:
```
mbedtls_ssl_tls13_write_record_size_limit_ext(mbedtls_ssl_context *ssl,
                                                  unsigned char *buf,
                                                  const unsigned char *end,
                                                  size_t *out_len)
```
2) To remove the call to `mbedtls_ssl_is_record_size_limit_valid()`
3) To use MBEDTLS_SSL_IN_CONTENT_LEN as the record size limit and to add here a static assert that
MBEDTLS_SSL_IN_CONTENT_LEN is greater than MBEDTLS_SSL_RECORD_SIZE_LIMIT_MIN.